opnode/p2p/gossip.go

```diff
@@ -0,0 +1,403 @@
+		var topicLen [8]byte
+		binary.LittleEndian.PutUint64(topicLen[:], uint64(len(topic)))
+		h.Write(topicLen[:])
+		h.Write([]byte(topic))
+		h.Write(data)
+		// the message ID is shortened to save space, a lot of these may be gossiped.
+		return string(h.Sum(nil)[:20])
+	}
+}
+
+func BuildGlobalGossipParams(cfg *rollup.Config) pubsub.GossipSubParams {
+	params := pubsub.DefaultGossipSubParams()
+	params.D = 8                                      // topic stable mesh target count
+	params.Dlo = 6                                    // topic stable mesh low watermark
+	params.Dhi = 12                                   // topic stable mesh high watermark
+	params.Dlazy = 6                                  // gossip target
+	params.HeartbeatInterval = 500 * time.Millisecond // frequency of heartbeat, seconds
+	params.FanoutTTL = 24 * time.Second               // ttl for fanout maps for topics we are not subscribed to but have published to, seconds
+	params.HistoryLength = 12                         // number of windows to retain full messages in cache for IWANT responses
+	params.HistoryGossip = 3                          // number of windows to gossip about
+
+	return params
+}
+
+func NewGossipSub(p2pCtx context.Context, h host.Host, cfg *rollup.Config) (*pubsub.PubSub, error) {
+	denyList, err := pubsub.NewTimeCachedBlacklist(30 * time.Second)
+	if err != nil {
+		return nil, err
+	}
+	return pubsub.NewGossipSub(p2pCtx, h,
+		pubsub.WithMaxMessageSize(maxGossipSize),
+		pubsub.WithMessageIdFn(BuildMsgIdFn(cfg)),
+		pubsub.WithNoAuthor(),
+		pubsub.WithMessageSignaturePolicy(pubsub.StrictNoSign),
+		pubsub.WithSubscriptionFilter(BuildSubscriptionFilter(cfg)),
+		pubsub.WithValidateQueueSize(maxValidateQueue),
+		pubsub.WithPeerOutboundQueueSize(maxOutboundQueue),
+		pubsub.WithValidateThrottle(globalValidateThrottle),
+		pubsub.WithPeerExchange(false),
+		pubsub.WithBlacklist(denyList),
+		pubsub.WithGossipSubParams(BuildGlobalGossipParams(cfg)),
+	)
+	// TODO: pubsub.WithPeerScoreInspect(inspect, InspectInterval) to update peerstore scores with gossip scores
+}
+
+func validationResultString(v pubsub.ValidationResult) string {
+	switch v {
+	case pubsub.ValidationAccept:
+		return "ACCEPT"
+	case pubsub.ValidationIgnore:
+		return "IGNORE"
+	case pubsub.ValidationReject:
+		return "REJECT"
+	default:
+		return fmt.Sprintf("UNKNOWN_%d", v)
+	}
+}
+
+func logValidationResult(self peer.ID, msg string, log log.Logger, fn pubsub.ValidatorEx) pubsub.ValidatorEx {
+	return func(ctx context.Context, id peer.ID, message *pubsub.Message) pubsub.ValidationResult {
+		res := fn(ctx, id, message)
+		var src interface{}
+		src = id
+		if id == self {
+			src = "self"
+		}
+		log.Debug(msg, "result", validationResultString(res), "from", src)
+		return res
+	}
+}
+
+func BuildBlocksValidator(log log.Logger, cfg *rollup.Config) pubsub.ValidatorEx {
+
+	// Seen block hashes per block height
+	// uint64 -> []common.Hash
+	blockHeightLRU, err := lru.New(100)
```

Comment:
Is this LRU shared among all peers, or just one? If it's shared among multiple peers, isn't there a race condition in which multiple peers could try and append to the `seen` slice at the same time? Or does the LRU copy it on the way out?

Reply:
Good catch, the LRU is shared and can be concurrently called by at most 4 validator routines, so there's a race condition indeed. See [0fca724](https://github.com/ethereum-optimism/optimistic-specs/pull/394/commits/0fca7248b28f90350bfe99d45ab3ae3e4fbc72ed) with fix + related misc improvements.